### ISSUE TITLE
add decompose function

### ISSUE THREAD

**Issue by chrvadala** [chrvadala/transformation-matrix#2]

**Comment by chrvadala** [comment]
Hi @peebeebee, do you have some progress on this?

**Comment by peebeebee** [comment]
No, I’d normally do it during my work hours, since it’s work related, but we’ve found a way around the problem sadly. 
Still on my to-do list though!

**Comment by chrvadala** [comment]
you’re right, are you open to submit a PR?

**Comment by chrvadala** [comment]
https://github.com/WebKit/webkit/blob/ef68f7e8a9eb4f83d68f69aca80f3a52ea0eecb0/Source/WebCore/platform/graphics/transforms/AffineTransform.cpp#L397

**Comment by chrvadala** [comment]
That is fantastic! It should close this old issue 🙂
Are you available to submit a PR?
Do you have some references (book, link, algorithm name,...) about this algorithm?

**Comment by hillin** [comment]
I think we've referenced several posts on Stack Overflow and Mathematics Stack Exchange, and a few other implementations in other libraries, with some homebrew (mostly the flip removing part).

I'd be happy to submit a PR but what's the code of conduct of tests here? We didn't write any test for this function.

**Comment by chrvadala** [comment]
I know, this algorithm is really tricky and it is important that it covers all the cases. This is why I asked you about some algorithm references. Unfortunately there are some algorithm that covers just a subset of transformation matrices.
Anyway, let's understand it with testing.

Regarding the tests I think that it will be really useful to create some matrices (leveraging on the other functions available in this library) and verify that they are "decomposable". We have to test both primitive matrices (e.g. translate) and composed matrices (e.g. translate+scale+rotate)

examples
````javascript
const matrix = translate(42, 42)

expect(decompose(matrix)).toEqual({
        translate: { tx: 20, y: 20 },
        rotation: 0,
        scale: { x: 1, y: 1 }
})
````

Just one note about the returned object. In order to keep consistency with https://github.com/chrvadala/transformation-matrix#fromdefinitiondefinitionorarrayofdefinition--arraymatrix signature, can I suggest the following alternative signature? What do you think?
```json
{
        translate: { tx: 0, ty: 0 },
        rotation: { angle: 0 },
        scale: { sx: 1, sy: 1 }
}
```

**Comment by hillin** [comment]
I‘m totally OK with the alternative signature, in our case we are just reusing some existing structs.

I'll see what I can do with then PR when I can catch some time.

**Comment by hillin** [comment]
I've created PR #88.
I find it a little bit disorienting when using the `flipX` and `flipY` method, which I instinctively thought stands for horizontal flip and vertical flip respectively, while the truth is contrary. Anyways I've adapted the code to this naming convention.

**Comment by chrvadala** [comment]
Released with v2.10.0
